File: day23/day23.kt
class Circle(input: String, val total: Int) {
    data class Node(val label: Int, var next: Node?)
    val nodes = List<Node>(total) { Node(it + 1, null) }

    init {
        val a = input.map { it - '1' }
        (0 until total).map { i ->
            val j = a.getOrElse(i) { it }
            val k = a.getOrElse(i + 1) { if (it < total) it else a[0] }
            nodes[j].next = nodes[k]
        }
    }
    
    fun slice(label: Int, count: Int): List<Int> {
        var node = nodes[label - 1]
        return (1..count).map { node = node.next!!; node.label }
    }
    
    fun move(label: Int, count: Int, block: Int = 3) {
        var node = nodes[label - 1]
        (1..count).forEach {
            var temp = node
            val slice = (0..block).map { temp = temp.next!!; temp }
            val labels = slice.dropLast(1).map { it.label }.toSet()
            var n = node.label
            do { n = if (n > 1) n - 1 else total } while (labels.contains(n))
            val dest = nodes[n - 1]
            slice[slice.size - 2].next = dest.next
            dest.next = slice[0]
            node.next = slice[slice.lastIndex]
            node = node.next!!
        }
    }
}

fun main() {
    val input = "137826495"

    val small = Circle(input, 9)
    small.move(1, 100)
    println(small.slice(1, 8).joinToString(""))

    val big = Circle(input, 1000000)
    big.move(1, 10000000)
    println(big.slice(1, 2).fold(1L) { a, b -> a * b })
}

File: 2020/day23/day23.py
input_ = "137826495"

class Node:
    def __init__(self, label):
        self.label = label
        self.next = None

class Circle:
    def __init__(self, initial, total):
        self.nodes = [Node(i + 1) for i in range(total)]
        m, a = len(initial), [int(c) - 1 for c in initial]
        for i, n in enumerate(map(int, initial)):
            p = a[i + 1] if i < m - 1 else (m if m < total else a[0])
            self.nodes[n - 1].next = self.nodes[p]
        for i in range(m, total):
            p = i + 1 if i < total - 1 else a[0]
            self.nodes[i].next = self.nodes[p]
        self.total = total

    def get(self, label, count):
        node, res = self.nodes[label - 1], []
        for _ in range(count):
            node = node.next
            res.append(node.label)
        return res

    def move(self, label, count, block=3):
        node = self.nodes[label - 1]
        for _ in range(count):
            temp, res = node, []
            for _ in range(block + 1):
                temp = temp.next
                res.append(temp)

            labels = [x.label for x in res[:-1]]
            n = node.label
            while True:
                n -= 1
                if not n: n = self.total
                if n not in labels: break
            dest = self.nodes[n - 1]

            res[-2].next = dest.next
            dest.next = res[0]
            node.next = res[-1]
            node = node.next

small = Circle(input_, 9)
small.move(1, 100)
print(''.join(map(str, small.get(1, 8))))

big = Circle(input_, 1000000)
big.move(1, 10000000)
a, b = big.get(1, 2)
print(a * b)
